### PR TITLE
Client-side deletion with callback

### PR DESCRIPTION
**Pull request by YasserHassan**
A feature to delete grid rows on the client side with the option to use a callback function for further processing on the client if required. Currently it should be used only in tabular forms as the deletion process would not be complete without form submission and execution on the server.

The callback could be used to collect the IDs of the deleted rows/records in a hidden input to submit them with the form to the server to execute the actual deletion from the database. Another option is to compare the submitted records with the ones in the database to identify which ones should be deleted.

The callback should not be used to attempt to delete the records using AJAX as there's no way to undelete a row if the call failed.

This feature aims at streamlining the user experience with grids and tabular forms. More features are expected later including:
- An optional pre-delete function that could be used to confirm deleting the record on the server via an AJAX call prior to deleting it on the client. This can extend the client-side deletion to grid views without form and deleting detail records in master-detail forms without form submission.
- Adding new rows to the grid. Typically the contents would be filled out in a dialog box then used to add the new row.

>Note: Grid changes need to be applied on the server to be preserved through grid page navigation.

### CODE REVIEW

**Comment by kartik-v** [comment]
Thanks for the efforts. Here are my suggestions:

- If this feature is only applicable for TabularForm then it must be a part of TabularForm implementation rather than GridView
- Trying to understand the use cases for the above... or why this is really necessary:
   - Client delete means you are trying to delay the server delete OR basically mark several records to delete as a batch. In this case options to
    - If someone needs to delete records in batches - one can make use of the CheckboxColumn to mark the rows for deletion and press a single delete button. Check [tabular form example scenario # 2](http://demos.krajee.com/builder-details/tabular-form#scenario-2) where an example of batch delete is shown. This maybe more intuitive than individually clicking delete for each row. You can easily control the batch delete action to purge only client data or server data based on identifying if the records are new.
    - Individually deleting each row. You can still control the behavior for this by trapping its on click event.

**Comment by YasserHassan** [comment]
Thank you @kartik-v for your evaluation, please find my answers below:

This feature is applicable to grid view too as well as tabular forms, just not safe (yet) for use outside of a form because AJAX calls could fail. But once a pre-delete function is introduced (could be very shortly), it should be safe to use it in grid views too like index pages. So the better place is in grid views.

A use case is a master-detail forms where the tabular form would be too big to fit in a row, a grid view could be used instead (saving unnecessary processing and traffic overhead) to update the records by opening a form in a dialog box from a row action button and update the record via AJAX. Deleted records from the grid would be saved to a hidden input to commit later when the form is submitted.

>Client delete means you are trying to delay the server delete OR basically mark several records to delete as a batch.

It is trying to identify several individual records for deletion as a batch in a delayed server action as follows:
1- Records are visually deleted from the grid view/tabular form (removed from HTML).
2- Deleted records could be identified on client-side (by the developer) by using the callback to gather deleted records' IDs in a hidden input (e.g. as comma-separated list) to submit with the form.
3- Actual deletion from the database is done when the form is submitted with or without the help of identified records above, depending on the developer's choice.

Future Enhancement: 4- Perform immediate deletion on server via AJAX then delete from the UI upon success.

Note: The above cases include delayed action scenarios that can not be performed using PJAX.

>If someone needs to delete records in batches - one can already currently make use of the CheckboxColumn to mark the rows for deletion and press a single delete button.

My understanding is the deletion in the given example currently requires page reload which I'm trying to avoid especially for master-detail forms. I agree that batch row deletion is more intuitive but only in batch delete. In fact there could be no contradiction between both if you look at this feature in part as an API to delete rows from the grid (kvDeleteRow). I'm thinking of extending kvDeleteRow to take an optional array of IDs.

>Individually deleting each row. You can still control the behavior for this by trapping its on click event.

I tried to do this first by handling the confirm event of Yii's `data-confirm` but turned out that there's no event triggered from there. I am already handling the on click event in a manner similar to Yii's `data-confirm` using `data-delete` to avoid conflict between both, please let me know if you have another opinion.

What if we structure this as separate grid action as follows:
1- kvDeleteRow: Grid view API to delete single or multiple rows passed as an array of IDs.
2- kvAddRow: API to add single or multiple rows based on a client-side template.
3.1- Add dedicated attributes to action buttons like `data-delete` and `data-edit` to handle client-side actions of each or...
3.2- Add general `data` attributes to action buttons like `data-event` and `data-action` as what event to handle and what to do with it, to handle a certain events like 'click', extending the possibilities of this area.

In points 3.1 and 3.2, the actions of AJAX-deleting a row or storing its ID, and the action of visually deleting that row could be decoupled, putting more control into the developer's hands.

Please let me know what you think. If you decide to merge this feature, please mark it as "Experimental with possible breaking updates" to share with the community until settled.

**Comment by YasserHassan** [comment]
Related to kartik-v/yii2-builder#71

**Comment by YasserHassan** [comment]
I was thinking @kartik-v, I am already using this feature via an extended class and submitted it here because it might become useful to other developers. But if you feel it is not mature enough yet, let's drop it for now and let me continue working on and come back later after it evolves and matures more.

**Comment by kartik-v** [comment]
IMO... better to use CheckboxColumn for most of this - you can select rows better, validate the selection, and then trigger the delete. You can also refresh the grid via PJAX after delete without needing to refresh page.

The advantage of using CheckboxColumn is you can define multiple batch action buttons for the same Column selection. For example, you can have BATCH DELETE, BATCH UPDATE, BATCH PRINT, BATCH DOWNLOAD all - using the CheckboxColumn. 

Also a variation like the RadioColumn is available which allows single row selection.

Using ActionColumn buttons you can individually modify or delete the row - its easy to override the button with javascript/jquery by triggering the onClick event.

**Comment by YasserHassan** [comment]
I need to clarify this @kartik-v. The actual benefit here is not single or batch actions, it is providing the ability to:
- Manipulate the grid view and tabular form rows by deleting and adding (later) individual rows, whether individual or batch. Consider it a grid view API. Deleting records on button click comes separately by using that API.
- Perform this manipulation in an offline mode (without referring to server) and submitting everything once for processing.
- Provide a better approach to master-detail forms especially in creating new master and detail records in one swap. In such case, there are no master or detail data in the database to use with PJAX. The new detail record could be submitted to the server with the whole form every time one is added to create the whole page and send it back to the browser and so on. This approach is not only more complicated on both client and server sides, but also the user experience would not be as streamlined.

For example, imagine creating a new invoice. You click a button to create a new invoice line, a new tabular form line is smoothly and instantly added to the existing lines to fill out. The same instant result when deleting individual or multiple invoice lines. Then all submitted once to the server for processing.

If you believe this is not a useful and coveted feature, let's drop it to avoid complicating the extension needlessly. Please let me know.

**Comment by kartik-v** [comment]
I think it would largely depend on the more important thing... how you add new rows from the client. That is more important and if you solve that first - this will be relatively trivial IMO...

**Comment by YasserHassan** [comment]
My approach to adding new form and grid records is still rough but it goes as follows:
1- Add a new grid view parameter like `'clientCreate'` => true` upon which a row template is automatically created as one of the following:
- A single grid/form template row is generated and saved in a JS variable properly escaped using a new PHP function like `generateRowTemplate` or...
- An extra hidden row is added as the first or last row to the generated grid and used as a template (even empty grids of course) and used as the template source. This approach more suitable for grid views than forms.
- A new hidden grid with a single row is added above or below the original grid and used as the template source.

2- Add another parameter like `templateProcessor` that takes a JS function name or anonymous function provided by the developer that takes the template HTML, prepares it, and returns the content to be inserted in the grid. He needs to make sure this function can processes a row from the grid (say the first one) and that the grid is prepared for that. Probably by using CSS classes for easy jQuery selectors.

3- If would be good if we can avoid using the sequence numbers in the tabular form elements without disturbing the order on the server-side, but if it's absolutely necessary, a new JS variable like `maxRowNumber` will be introduced to use for sequencing new rows and incremented with each one but not necessarily decremented on deletion. In this case, an template placeholder like {rowNumber} could be used for numbering. This is not the developer's work, his processing function would receive the template already sequenced.

Later enhancement:
4- For paginated grid views and forms, there could be an option to `escapeIfFull` where the record will be inserted only if the current page is the first or last there's still space. Pagination will need to be adjusted then but this is another story.

Notice that grid views can benefit from this too by creating a new record using a dialog box form, submit it for creation via AJAX, and add the new row on success. However, non-paginated tabular forms in master-detail view is the optimum situation.

**Comment by kartik-v** [comment]
Note - the rendering of each new row will not be trivial ... it needs to account for complex scenarios... there are more than one way to do it... but the safest way is to use a renderAjax method to optimally use the GridView's row rendering methods.

Reasons - 
- just using a template is a simpler use case... it would help only if it were plain simple HTML inputs to render for each row.
- however, the advanced scenarios would involve using input widgets and jquery plugins - these cannot be parsed easily via javascript. 
- we then cannot have a generic approach if the developer has to initialize every jquery plugin or widget on his own - he/she may need to write the entire javascript

Having said that, there is a way with krajee input widgets and plugins which has a standardized way so reinitializing via javascript will be pretty fast. So we need to combine both.

**Comment by YasserHassan** [comment]
I believe it's better to work on this over two phases. Phase 1 is restricted to simple plain HTML form elements. I believe this alone will address the majority of needs. Phase 2 is for the more complex scenarios. Even without phase 2, it would be a tradeoff between having advanced form elements or a smoother user experience. Personally I'd go for the latter.

renderAjax still could be used by the developer's processor function. It's up then to the developer to call the server and provide its result with a callback to initialize elements after rendering if required. For more complex scenarios, the developer's could ignore the API and handle the whole processes on his own.

**Comment by kartik-v** [comment]
Agree. But the expectation for developers will be to have this consistent across.

Currently TabularForm is an advanced tool that offers rendering HTML inputs, advanced widgets, and jquery plugin - based widgets and batch submit them.

So in the future scenario, we will have a tabular batch form with advanced widgets available to update records... but only plain HTML inputs available for creating. The added challenge for the developer is that the configuration of Tabular columns will not be consistent for create and update.

We need to study this --- but IMO... we need to think on maintaining the feature consistency and make it simpler and uniform for developers.

In its current form - using an ajax based create/update is the additional code needed for maintaining the user experience for master detail forms. A few additional enhancements need to be done to Yii's DataProvider (or creating a new DataProvider) to handle models in an array.

**Comment by YasserHassan** [comment]
I agree, consistency is important while we can have variations to how things are done; pure client creation via template(s) as well as server-assisted creation via renderAjax. It's like a toolbox for the developer to use from knowing the advantages and limitations of each approach.

In phase 1, simple HTML elements would be used for both creating and updating records. Typically all records would look the same. Even if different form elements are used in each row, all should remain simple HTML. This could be addressed as a known restriction in phase 1 and even could be checked during form creation to prevent using advanced elements with `clientCreate` and issue an error if so.

Yes it needs study and I'll keep working on my extended class to evolve it more and have a clearer idea about the best approaches we can use.

>In its current form - using an ajax based create/update is the additional code needed for maintaining the user experience for master detail forms.

Did you mean to create/update data or UI? If UI, creating new rows only is what we need. A special attention is required for creating master-detail data in one swap.

>A few additional enhancements need to be done to Yii's DataProvider (or creating a new DataProvider) to handle models in an array.

Wouldn't `ArrayDataProvider` be sufficient for that? `ActiveDataProvider` or a derivative would be better to benefit from the `ActiveForm` validation features. I have already developed something like that to use JSON data via model templates in ActiveDataProvider and it works well with TabularForm. That's why I asked in kartik-v/yii2-builder#71 about the side effects of having no record keys. It's still evolving but I can share it if required.

**Comment by kartik-v** [comment]
For your queries on my last two points - what I meant is without any change to the current code JUST in its current form - the options are already available to each developer to build their own UI for various scenarios with some additional javascript.

**Comment by kartik-v** [comment]
Also thinking loud... another solution... just in its current code format... 

- UPDATE: you can use `TabularForm` for all batch UPDATE.
- CREATE: you can popup a modal dialog and create a new record using `Form` builder. Then submitting the modal form can refresh your tabular form grid.

The above scenario is simple and can work across any scenario of complex layouts. It can also be useful when you have different form fields needed for CREATE and UPDATE.

That's another complexity when you want to create new records in a tabular form. The solution will only work if the number of fields and field layout exactly matches for both - else it becomes complex.

**Comment by kartik-v** [comment]
Also thinking loud... another solution... just in its current code format without any change... 

- UPDATE: you can use `TabularForm` for all batch UPDATE.
- CREATE: you can popup a modal dialog and create a new record using `Form` builder. Then submitting the modal form can refresh your tabular form grid.

The above scenario is simple and can work across any scenario of complex layouts. It can also be useful when you have different form fields needed for CREATE and another different set of fields needed for UPDATE. In the solution you are trying to build this will be not possible to manage (having different fields for create and different fields for update - since its bound by the tabular layout).

**Comment by YasserHassan** [comment]
I think we have tackled the top level details and we have a better understanding about this area. This discussion was fruitful. This commit will not provide the constancy required as it needs to factor in the points we discussed. I say you drop it while I keep working on these features via my extended classes and commit another one if things evolved nicely.

**Comment by kartik-v** [comment]
No problems - its useful we discuss - just did not want you to waste effort before thinking out the implementation outcomes. 

BTW... I may have a javascript extension to all the krajee widgets and plugins soon so that they can be used easily like any other HTML input.

**Comment by YasserHassan** [comment]
Thank you for your consideration @kartik-v, it's good to think ahead. It'd be interesting to see the new script, I'll keep watching. I'll cancel this PR.
